src/tpm2_wrap.c

```diff
@@ -1784,29 +1784,58 @@ int wolfTPM2_RsaKey_TpmToWolf(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
     return rc;
 }
 
-int wolfTPM2_RsaKey_TpmToPem(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
-                             byte* pem, int* pemSz, byte* tempBuf, int tempSz)
+int wolfTPM2_RsaKey_TpmToPemPub(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
+    byte* pem, word32* pemSz)
 {
     int rc = TPM_RC_FAILURE;
 #if !defined(WOLFTPM2_NO_WOLFCRYPT) && defined(WOLFSSL_DER_TO_PEM)
-    int derSz;
     RsaKey rsaKey;
+    byte* derBuf = NULL;
+    int derSz = 0;
+#endif
+
+    if (dev == NULL || tpmKey == NULL || pem == NULL || pemSz == NULL)
+        return BAD_FUNC_ARG;
 
+#if !defined(WOLFTPM2_NO_WOLFCRYPT) && defined(WOLFSSL_DER_TO_PEM)
     /* Prepare wolfCrypt key structure */
     rc = wc_InitRsaKey(&rsaKey, NULL);
-    if (rc != 0) return rc;
-    /* Convert the wolfTPM key to wolfCrypt format */
-    rc = wolfTPM2_RsaKey_TpmToWolf(dev, tpmKey, &rsaKey);
-    if (rc != 0) return rc;
-    /* Convert the wolfCrypt key to DER format */
-    rc = wc_RsaKeyToPublicDer(&rsaKey, tempBuf, tempSz);
-    if (rc <= 0) return rc;
-    derSz = rc;
-    /* Convert the DER key to PEM format */
-    rc = wc_DerToPem(tempBuf, derSz, pem, tempSz, PUBLICKEY_TYPE);
-    if (rc <= 0) return rc;
-    *pemSz = rc;
-    rc = TPM_RC_SUCCESS;
+    if (rc == 0) {
+        /* Convert the wolfTPM key to wolfCrypt format */
+        rc = wolfTPM2_RsaKey_TpmToWolf(dev, tpmKey, &rsaKey);
+        if (rc == 0) {
+            /* Get DER size - newer API can be called with NULL to get size */
+            rc = wc_RsaKeyToPublicDer(&rsaKey, NULL, 0);
+            if (rc > 0) {
+                derSz = rc;
+            }
+            else if (rc == BAD_FUNC_ARG) {
+                /* for older wolfSSL estimate based on key size */
+                derSz = wc_RsaEncryptSize(&rsaKey) * 2;
+                rc = 0;
+            }
+        }
+        if (derSz > 0) {
+            derBuf = (byte*)XMALLOC(derSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
I think this is the first ever XMALLOC in wolfTPM. It is in the examples, but are we okay with starting to use HEAP?

We will have to update the product page.

Reply:
It is not the first. Since this already requires wolfCrypt and the PEM code is already doing malloc it is fine.

Reply:
you are right. Good for merge then :)